File: app/src/main/java/com/example/doankotlin/Activity/CartActivity.java
package com.example.doankotlin.Activity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.doankotlin.Adapter.CartAdapter;
import com.example.doankotlin.Domain.Foods;
import com.example.doankotlin.Helper.ManagmentCart;
import com.example.doankotlin.databinding.ActivityCartBinding;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class CartActivity extends BaseActivity {

    private ActivityCartBinding binding;
    private RecyclerView.Adapter adapter;
    private ManagmentCart managmentCart;
    private double tax;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityCartBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        managmentCart = new ManagmentCart(this);

        setVariable();
        calculateCart();
        initList();
    }

    private void initList() {
        if(managmentCart.getListCart("CartList").isEmpty()){
            binding.emptyTxt.setVisibility(View.VISIBLE);
            binding.scrollviewCart.setVisibility(View.GONE);
        }else {
            binding.emptyTxt.setVisibility(View.GONE);
            binding.scrollviewCart.setVisibility(View.VISIBLE);
        }

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this,LinearLayoutManager.VERTICAL,false);
        binding.cardView.setLayoutManager(linearLayoutManager);
        adapter = new CartAdapter(managmentCart.getListCart("CartList"), this, this::calculateCart);

        binding.cardView.setAdapter(adapter);
    }

    private void calculateCart() {
        double percentTax = 0.04; // percent 4%
        double delivery = 1000; // 1000 VND

        tax = Math.round(managmentCart.getTotalFee("CartList")*percentTax*100.0)/100;

        double total = Math.round((managmentCart.getTotalFee("CartList")+tax+delivery)*100)/100;
        double itemTotal = Math.round(managmentCart.getTotalFee("CartList")*100)/100;

        binding.totalFeeTxt.setText(itemTotal +"đ");
        binding.taxTxt.setText(tax+"đ");
        binding.deliveryTxt.setText(delivery+"đ");
        binding.totalTxt.setText(total + "đ");

    }

    private void setVariable() {
        binding.backBtn.setOnClickListener(v -> finish());
        binding.orderBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ArrayList<Foods> orderList = managmentCart.getListCart("CartList");
                FirebaseUser user = mAuth.getCurrentUser();

                Map<String, Object> foodInfo = new HashMap<>();
                for(int i=0; i<orderList.size();i++) {
                    DatabaseReference myRef = database.getReference("Orders").child(user.getUid()).child(String.valueOf(i));
                    Foods food = orderList.get(i);
                    //myRef.child(String.valueOf(food.getId()));
                    foodInfo.put("Title", food.getTitle());
                    foodInfo.put("Quantity", food.getNumberInCart());
                    foodInfo.put("Total", food.getPrice() * food.getNumberInCart());
                    foodInfo.put("ImagePath", food.getImagePath());

                    myRef.setValue(foodInfo);
                    Toast.makeText(CartActivity.this,"Đặt hàng thành công", Toast.LENGTH_SHORT).show();
                    Log.d("TAG","Data: " + foodInfo);
                }
            }
        });
    }
}

File: app/src/main/java/com/example/doankotlin/Activity/CartActivity.kt
package com.example.doankotlin.Activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.doankotlin.Adapter.CartAdapter
import com.example.doankotlin.Domain.CartItems
import com.example.doankotlin.Domain.Foods
import com.example.doankotlin.databinding.ActivityCartBinding
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import kotlin.math.roundToInt

class CartActivity : BaseActivity() {
    private lateinit var binding: ActivityCartBinding
    private var tax = 0.0
    private lateinit var cartAdapter: CartAdapter
    private lateinit var userId: String
    private var receivedData = 0.0
    private var num = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCartBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setVariables()
        retrieveCartItems()
    }

    private fun retrieveCartItems() {
        userId = mAuth.currentUser?.uid ?: ""
        val cartReference: DatabaseReference =
            database.reference.child("Users").child(userId).child("CartItems")

        val cartItemList: ArrayList<Foods> = ArrayList()

        cartReference.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                cartItemList.clear()
                receivedData = 0.0 // Reset receivedData
                num = 0 // Reset num
                for (cartSnapshot in snapshot.children) {
                    val cartItems = cartSnapshot.getValue(CartItems::class.java)
                    cartItems?.let {
                        val foodItem = Foods().apply {
                            title = it.Title
                            price = it.Price ?: 0
                            imagePath = it.ImagePath
                            numberInCart = it.Quantity ?: 0
                        }
                        cartItemList.add(foodItem)
                        receivedData += foodItem.price * foodItem.numberInCart
                        num += foodItem.numberInCart
                    }
                }
                if(cartItemList.isEmpty()){
                    binding.emptyTxt.visibility = View.VISIBLE
                    binding.scrollviewCart.visibility = View.GONE
                } else {
                    binding.emptyTxt.visibility = View.GONE
                    binding.scrollviewCart.visibility = View.VISIBLE
                }
                setAdapter(cartItemList)
            }

            override fun onCancelled(error: DatabaseError) {
                Toast.makeText(this@CartActivity, "Error retrieving cart items", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun setAdapter(cartItems: ArrayList<Foods>) {
        cartAdapter = CartAdapter(cartItems, this@CartActivity) {
            calculateCart()
        }
        binding.cardView.layoutManager = LinearLayoutManager(this@CartActivity)
        binding.cardView.adapter = cartAdapter

        // Calculate cart total initially
        calculateCart()
    }

    private fun calculateCart() {
        var delivery = 5000.0 // 5000 VND
        var percentTax = 0.05 * (num/5)

        if(percentTax >= 0.3) percentTax = 0.3
        if(receivedData >= 200000) delivery = 0.0

        tax = (Math.round(receivedData * percentTax * 100.0) / 100).toDouble()
        val total = (Math.round((receivedData - tax + delivery) * 100) / 100).toDouble()
        val itemTotal = (Math.round(receivedData * 100) / 100).toDouble()

        binding.totalFeeTxt.text = "${itemTotal}đ"
        binding.taxTxt.text = "${tax}đ"
        binding.deliveryTxt.text = "${delivery}đ"
        binding.totalTxt.text = "${total}đ"
    }

    private fun setVariables() {
        binding.backBtn.setOnClickListener { finish() }
        binding.orderBtn.setOnClickListener {
            // Directly pass cart ItemList to orderNow function
            orderNow(cartAdapter.getCartItems())
        }
    }

    private fun orderNow(listFood: ArrayList<Foods>) {
        val intent = Intent(this@CartActivity, PayOutActivity::class.java)
        intent.putExtra("listFood", listFood)
        startActivity(intent)
    }
}